uaclient/apt_news.py

```diff
@@ -0,0 +1,153 @@
+from uaclient.config import UAConfig
+from uaclient.data_types import (
+    BoolDataValue,
+    DataObject,
+    DatetimeDataValue,
+    Field,
+    StringDataValue,
+    data_list,
+)
+from uaclient.files import state_files
+
+
+# TODO: list of selectors is undecided
+class AptNewsMessageSelectors(DataObject):
+    fields = [
+        Field("pro", BoolDataValue, required=False),
+    ]
+
+    def __init__(self, pro: bool):
+        self.pro = pro
+
+
+class AptNewsMessage(DataObject):
+    fields = [
+        Field("begin", DatetimeDataValue),
+        Field("end", DatetimeDataValue, required=False),
+        Field("selectors", AptNewsMessageSelectors, required=False),
+        Field("lines", data_list(StringDataValue)),
+    ]
+
+    def __init__(
+        self,
+        begin: datetime.datetime,
+        end: Optional[datetime.datetime],
+        selectors: Optional[AptNewsMessageSelectors],
+        lines: List[str],
+    ):
+        self.begin = begin
+        self.end = end
+        self.selectors = selectors
+        self.lines = lines
+
+
+def is_control_char(c: str) -> bool:
+    return unicodedata.category(c)[0] == "C"
+
+
+def clean_line(line: str) -> str:
+    line_without_control_chars = "".join(
+        [c for c in line if not is_control_char(c)]
+    )
+    stripped_line = line_without_control_chars.strip()
+    return stripped_line
+
+
+def format_message(msg: AptNewsMessage) -> str:
+    result = "#\n"
+    for line in msg.lines:
+        cleaned_line = clean_line(line)
+        if len(cleaned_line) > 0:
+            result += "# {}\n".format(cleaned_line)
+    result += "#\n"
+    return result
+
+
+# TODO: list of selectors is undecided
+def do_selectors_apply(selectors: Optional[AptNewsMessageSelectors]) -> bool:
+    if selectors is None:
+        return True
+    return True
+
+
+def do_dates_apply(
+    begin: datetime.datetime, end: Optional[datetime.datetime]
+) -> bool:
+    now = datetime.datetime.now(datetime.timezone.utc)
+    if now < begin:
+        return False
+
+    one_month_after_begin = begin + datetime.timedelta(days=30)
+    if end is None or end > one_month_after_begin:
+        end_to_use = one_month_after_begin
+    else:
+        end_to_use = end
+    if now > end_to_use:
+        return False
+
+    return True
+
+
+def is_message_valid(msg: AptNewsMessage) -> bool:
+    if len(msg.lines) < 1:
+        return False
+    if len(msg.lines) > 3:
+        return False
+
+    for line in msg.lines:
+        if "\n" in line or "\r" in line:
+            # All control characters will be removed.
+            # Newlines are particularly important to formatting expectations.
+            # Since we can't honor newlines, and removing them could
+            # substantially change meaning, we consider messages with them
+            # to be invalid.
+            return False
+        if len(line) > 77:
+            return False
+
+    if all(len(clean_line(line)) == 0 for line in msg.lines):
```

Comment:
What happens if one of the lines after cleaning has a zero length size ? Can it potentially change the meaning of the message ?

I know we are already checking for the existence of newlines and tabs, but I am just wondering if there are other special character that could compromise the message if excluded

Reply:
From conversation: We'll just disallow control characters entirely instead of trying to remove them. I'll make that change

Reply:
done!